### PR TITLE
fix(express): fix wrong rpcMetadata.route value with router.use(<midd…

### PR DESCRIPTION
**Pull request by biuboombiuboom**
### What version of OpenTelemetry are you using?

    opentelemetry: 1.4.0
    node: 16.14.2

### What did you do?
        const app = express();
        const router = express.Router();
        app.use('/users', router);
        router.get('/:userId', (req, res) => {
          res.status(200).end('user-'+req.params.userId);
        });
        router.use(customMiddleware);
        const bookRouter = express.Router({ mergeParams: true });
        router.use('/:userId/books', bookRouter);
        bookRouter.get('/:bookId', (req, res) => {
          setImmediate(() => {
            res.status(200).end('book-' + req.params.bookId);
          });
        });       

    request http://localhost:8080/proxy/v1/bar/1

### What did you expect to see?

    rpcMetadata.route='/users/:userId/books/:bookId'

### What did you see instead?

    rpcMetadata.route='/:userId/books/:bookId'

### CODE REVIEW

**Comment by chigia001** [comment]
@biuboombiuboom can you try remove this condition in https://github.com/open-telemetry/opentelemetry-js-contrib/blob/82267ab02f8d9b27613b5926089c42d04d4a4a7e/plugins/node/opentelemetry-instrumentation-express/src/instrumentation.ts#L201
to see if it still pass all your test?

File: plugins/node/opentelemetry-instrumentation-express/src/instrumentation.ts [hunk 1]
```diff
@@ -269,7 +269,7 @@ export class ExpressInstrumentation extends InstrumentationBase<
               req.res?.removeListener('finish', onResponseFinish);
               span.end();
             }
-            if (!(req.route && arguments[0] instanceof Error)) {
+            if (!(req.route && arguments[0] instanceof Error) && layerPath) {
```
Comment (chigia001) [inline plugins/node/opentelemetry-instrumentation-express/src/instrumentation.ts:272]: so we are over pop item right,

If layerPath is undefined, we don't append new record _LAYERS_STORE_PROPERTY. But when next is call, in the current implementation we still pop it.

File: plugins/node/opentelemetry-instrumentation-express/test/express.test.ts [hunk 2]
```diff
@@ -275,6 +275,83 @@ describe('ExpressInstrumentation', () => {
...
+            finishListenerCount = res.listenerCount('finish');
+          });
+          next();
+        });
+        const router = express.Router();
+        app.use('/users', router);
+        router.get('/:userId', (req, res) => {
+          res.status(200).end('user-' + req.params.userId);
+        });
+        router.use(customMiddleware);
```
Comment (chigia001) [inline plugins/node/opentelemetry-instrumentation-express/test/express.test.ts:303]: This customMiddleware layer is the one that make we over remove one layer.

**Review by chigia001 — COMMENTED** [review]
I think we should simlify the test.

**Comment by chigia001** [comment]
@biuboombiuboom this logic potential can lead to the same issue right?
https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1615/files#diff-81c3e13c3dac53e0dd6aabf6dc165305350eb522a85a2561f9ee51d9249011e7L209

**Comment by chigia001** [comment]
@biuboombiuboom I think we can refactor storeLayerPath to make it more strict on push/pop action. 
```typescript
export const storeLayerPath = (request: PatchedRequest, value?: string): {route: string, popLayerPath: () => void} => {
  let requestRouteStack: string[]
  if (Array.isArray(request[_LAYERS_STORE_PROPERTY]) === false) {
    requestRouteStack = [];
    Object.defineProperty(request, _LAYERS_STORE_PROPERTY, {
      enumerable: false,
      value: requestRouteStack,
    });
  } else {
    requestRouteStack = request[_LAYERS_STORE_PROPERTY] as string[]
  }
  if (value !== undefined) {
    requestRouteStack.push(value)
  }

  const stackLength = requestRouteStack.length

  const route = requestRouteStack
          .filter(path => path !== '/' && path !== '/*')
          .join('');
  let alreadyPopStack = false
  return {
    route,
    popLayerPath: () => {
      if (value !== undefined) {
        return
      }
      if (requestRouteStack.length === stackLength && alreadyPopStack === false) {
        requestRouteStack.pop()
        alreadyPopStack = true
      } else {
        diag.error(
          'express instrumentation: pop route stack incorrectly'
        );
      }
    }
  }
};
```

**Comment by biuboombiuboom** [comment]
> @biuboombiuboom I think we can refactor storeLayerPath to make it more strict on push/pop action. It also prevent instrumentation.ts from knowing of this symbol property.
> 
> ```ts
> export const storeLayerPath = (request: PatchedRequest, value?: string): {route: string, popLayerPath: () => void} => {
>   let requestRouteStack: string[]
>   if (Array.isArray(request[_LAYERS_STORE_PROPERTY]) === false) {
>     requestRouteStack = [];
>     Object.defineProperty(request, _LAYERS_STORE_PROPERTY, {
>       enumerable: false,
>       value: requestRouteStack,
>     });
>   } else {
>     requestRouteStack = request[_LAYERS_STORE_PROPERTY] as string[]
>   }
>   if (value !== undefined) {
>     requestRouteStack.push(value)
>   }
> 
>   const stackLength = requestRouteStack.length
> 
>   const route = requestRouteStack
>           .filter(path => path !== '/' && path !== '/*')
>           .join('');
>   let alreadyPopStack = false
>   return {
>     route,
>     popLayerPath: () => {
>       if (value !== undefined) {
>         return
>       }
>       if (requestRouteStack.length === stackLength && alreadyPopStack === false) {
>         requestRouteStack.pop()
>         alreadyPopStack = true
>       } else {
>         diag.error(
>           'express instrumentation: pop route stack incorrectly'
>         );
>       }
>     }
>   }
> };
> ```
greate.

**Comment by biuboombiuboom** [comment]
> @biuboombiuboom I think we can refactor storeLayerPath to make it more strict on push/pop action. It also prevent instrumentation.ts from knowing of this symbol property.
> 
> ```ts
> export const storeLayerPath = (request: PatchedRequest, value?: string): {route: string, popLayerPath: () => void} => {
>   let requestRouteStack: string[]
>   if (Array.isArray(request[_LAYERS_STORE_PROPERTY]) === false) {
>     requestRouteStack = [];
>     Object.defineProperty(request, _LAYERS_STORE_PROPERTY, {
>       enumerable: false,
>       value: requestRouteStack,
>     });
>   } else {
>     requestRouteStack = request[_LAYERS_STORE_PROPERTY] as string[]
>   }
>   if (value !== undefined) {
>     requestRouteStack.push(value)
>   }
> 
>   const stackLength = requestRouteStack.length
> 
>   const route = requestRouteStack
>           .filter(path => path !== '/' && path !== '/*')
>           .join('');
>   let alreadyPopStack = false
>   return {
>     route,
>     popLayerPath: () => {
>       if (value !== undefined) {
>         return
>       }
>       if (requestRouteStack.length === stackLength && alreadyPopStack === false) {
>         requestRouteStack.pop()
>         alreadyPopStack = true
>       } else {
>         diag.error(
>           'express instrumentation: pop route stack incorrectly'
>         );
>       }
>     }
>   }
> };
> ```
greate.

**Comment by blumamir** [comment]
@biuboombiuboom Let me know when this is ready for review

**Pull request closed without merging**